doc/cascadia/Cascading-Default-Settings.md

```diff
@@ -0,0 +1,571 @@
+---
+author: Mike Griese @zadjii-msft
+created on: 2019-05-31
+last updated: 2019-07-31
+issue id: 754
+---
+
+# Cascading Default + User Settings
+
+## Abstract
+
+This spec outlines adding support for a cascading settings model. In this model,
+there are two settings files, instead of one.
+
+1. The default settings file
+2. The user's settings file
+
+The default settings file would be a static, read-only file shipped with the
+terminal. The user settings file would then contain all the user's chosen
+customizations to the settings. These two files would then be composed together
+when the app is launched, so that the runtime settings are the union of both the
+defaults and whatever modifications the user has chosen. This will enable the
+app to always use a default schema that it knows will be valid, and minimize the
+settings that the user needs to customize.
+
+Should the settings schema ever change, the defaults file will change, without
+needing to re-write the user's settings file.
+
+It also outlines a mechanism by which profiles could be dynamically added or
+hidden from the profiles list, based on some external source.
+
+## Inspiration
+
+Largely inspired by the settings model that both VS Code (and Sublime Text) use.
+
+### Goal: Minimize Writing to `profiles.json`
+
+We want to write the user settings file, `profiles.json`, as little as possible.
+Each time we write the file, there's the possiblity that we've re-ordered the
+keys, as `jsoncpp` provides no ordering guarantee of the keys. This isn't great,

```

Comment:
This sounds like a problem with the library we've chosen for json more than something desirable to design around.

Reply:
That's just one of the problems (and not one I'm totally sure _any_ json library supports tbh).

For example, when you have `"acrylicOpacity": 0.9`, we'll re-serialize that as `0.8999146garbage`, and that'll be enough for us to think the json changed. In general, we just want to not be re-serializing their json.

Reply:
any reason why we wouldn't just change that one to be "acrylicOpacity": "90%" and then store it as an integer instead of a float to prevent this? (adding appropriate parsers/converters)

I think we could work this out.